include/deal.II/lac/petsc_compatibility.h

```diff
@@ -138,6 +140,49 @@ namespace PETScWrappers
   {
     set_matrix_option(matrix, MAT_KEEP_NONZERO_PATTERN, PETSC_TRUE);
   }
+
+
+  /**
+   * Tell PETSc that the status of the vector has changed.
+   */
+  void
+  petsc_object_state_increase(Vec v);
+
+  /**
+   * Tell PETSc that the status of the matrix has changed.
+   */
+  void
+  petsc_object_state_increase(Mat A);
```

Comment:
Ah, interesting. I read up on it after I reviewed your patch, and if I understand correctly, the "state" is simply a counter one can increment (or set to anything) every time an object has been changed in a meaningful way (say, every time a matrix object has been assembled). 

Can *not* incrementing the state lead to bugs?

As for the name of the function, would you be willing to change it to `petsc_object_increment_state()` or `petsc_increment_state_counter()`? "Increase" is not obviously a verb to readers who don't know the concept, and a function name should contain a verb.

Reply:
> Can _not_ incrementing the state lead to bugs?

Yes because for example, we cache vector norms, that are invalidated if the state is increased. That's why now every loop on setValues call must be ended by assemblyBegin/End , otherwise we raise an error.

For matrices associated to preconditioners (i.e. KSPSetOperators/PCSetOperators), an increase in the state means that a preconditioner will be rebuilt

Reply:
Ah, awkward. I wonder how many bugs we then have. Do all of our `compress()` calls now increment the state counter? That might be worth auditing.

Reply:
I would say the deal.II wrappers are ok now, but I can double-check.

`compress()` calls assembly routines, which update the state counter internally. Also, I always use a debug version of PETSc that has extra checks for these situations. I caught some bugs in your tests while working on this stuff and already fixed them in old PRs. It might be worth using a debug version of PETSc in your CI runs.

Reply:
Interesting. I don't know much about the CI, but @tamiko or @marcfehling might be interested in this.